mono/utils/mono-safe-handle.c

```diff
@@ -0,0 +1,167 @@
+/**
+ * \file
+ * Support for synchronously freeing unmanaged memory subject to
+ * concurrent refcount modification
+ *
+ * Author:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * (C) 2018 Microsoft, Inc.
+ *
+ */
+
+#include <config.h>
+#include <glib.h>
+#include "utils/mono-safe-handle.h"
+#include "utils/atomic.h"
+
+#define INTERNAL_HANDLE_TOMBSTONE (GINT_TO_POINTER (-1))
+
+static void 
+ensure_synch_cs_set (MonoSafeHandleNative *handle)
+{
+	MonoCoopMutex *synch_cs;
+
+	if (handle->synch_cs != NULL) {
+		return;
+	}
+
+	synch_cs = g_new0 (MonoCoopMutex, 1);
+	mono_coop_mutex_init_recursive (synch_cs);
+
+	gpointer ret = mono_atomic_cas_ptr ((gpointer *)&handle->synch_cs, synch_cs, NULL);
+
+	if (ret != NULL) {
+		/* Another handle must have installed this CS */
+		/* Or the finalizer had to have run */
+		mono_coop_mutex_destroy (synch_cs);
+		g_free (synch_cs);
+	}
+}
+
+static void
+internal_handle_free (MonoSafeHandleNative *handle)
+{
+	MonoCoopMutex *synch_cs = handle->synch_cs;
+	handle->synch_cs = INTERNAL_HANDLE_TOMBSTONE;
+	mono_memory_barrier ();
+
+	mono_coop_mutex_destroy (synch_cs);
+	g_free (synch_cs);
+}
+
+static void
+change_wait_queue (MonoSafeHandleNative *handle, int diff)
+{
+	g_assert (diff == 1 || diff == -1);
+
+	mono_memory_barrier ();
+	MonoSafeHandleNativeState flags;
+	flags.mem_start = mono_atomic_load_i64 (&handle->lock_flags.mem_start);
```

Comment:
The runtime cannot use 64bits cas.

Reply:
That was my point -- gsize/ptr should be ok though, right? i.e. 64bit is ok on 64bit, right? 32bit on any.